connaisseur/config.py

```diff
@@ -0,0 +1,110 @@
+import json
+import yaml
+from jsonschema import validate, ValidationError
+from connaisseur.exceptions import NotFoundException, InvalidFormatException
+from connaisseur.util import safe_path_exists
+
+
+class Config:
+    CONFIG_PATH = "/etc/connaisseur-config/config.yaml"
+    CONFIG_SCHEMA_PATH = "/app/connaisseur/res/config_schema.json"
+    notaries: list = []
+
+    def __init__(self):
+        with open(self.CONFIG_PATH, "r") as configfile:
+            config_content = yaml.safe_load(configfile)
+
+        if not config_content:
+            raise NotFoundException("error getting any notary host configurations.")
+
+        with open(self.CONFIG_SCHEMA_PATH, "r") as schema_file:
+            schema = json.load(schema_file)
+
+        try:
+            validate(instance=config_content, schema=schema)
+        except ValidationError:
+            raise InvalidFormatException(
+                "invalid format for Connaisseur configuration."
+            )
+
+        self.notaries = [Notary(notary) for notary in config_content]
+
+    def get_notary(self, notary_name: str = None):
+        try:
+            if notary_name:
+                return next(
+                    notary for notary in self.notaries if notary.name == notary_name
+                )
+            return next(iter(self.notaries))
+        except StopIteration:
+            raise NotFoundException(
+                "the given notary configuration could not be found.",
+                {"notary_name": notary_name},
+            )
+
+
+class Notary:  # pylint: disable=too-many-instance-attributes
+    name: str
+    host: str
+    root_keys: list
+    has_auth: bool = False
+    auth: dict = {}
+    is_selfsigned: bool = False
+    selfsigned_cert: str = None
+    is_acr: bool = False
+
+    SELFSIGNED_PATH = "/etc/certs/{}.crt"
+    AUTH_PATH = "/etc/creds/{}/cred.yaml"
+
+    def __init__(self, notary_config: dict):
+        self.name = notary_config.get("name")
+        self.host = notary_config.get("host")
+        self.root_keys = notary_config.get("rootKeys")
+        self.is_acr = notary_config.get("isAcr", False)
+
+        if not (self.name and self.host and self.root_keys):
+            raise InvalidFormatException(
+                "error parsing the the Connaisseur configuration file."
+            )
+
+        self.is_selfsigned = safe_path_exists(
+            "/etc/certs/", self.SELFSIGNED_PATH.format(self.name)
+        )
+        self.has_auth = safe_path_exists(
+            "/etc/creds/", self.AUTH_PATH.format(self.name)
+        )
+
+    def get_key(self, key_name: str = None):
+        try:
+            if key_name:
+                key = next(
+                    key.get("key")
+                    for key in self.root_keys
+                    if key.get("name") == key_name
+                )
+            else:
+                key = next(iter(self.root_keys)).get("key")
+            return "".join(key.split("\n")[1:-2])
```

Comment:
What do the numbers in `[1:-2]` depend on (especially the `2`)? (Should we be trimming whitespace, for example?)

Reply:
this is a yaml snippet that gets transformed into json:

```yaml
root_keys:
  - name: library
    key: |
      -----BEGIN PUBLIC KEY-----
      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEOXYta5TgdCwXTCnLU09W5T4M4r9f
      QQrqJuADP6U7g5r9ICgPSmZuRHP/1AYUfOQW3baveKsT969EfELKj1lfCA==
      -----END PUBLIC KEY-----
```

here the json part of this:

```json
{
  "root_keys": [
    {
      "name": "library",
      "key": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEOXYta5TgdCwXTCnLU09W5T4M4r9f\nQQrqJuADP6U7g5r9ICgPSmZuRHP/1AYUfOQW3baveKsT969EfELKj1lfCA==\n-----END PUBLIC KEY-----\n"
    }
  ]
}
```

the `[1:-2]` will get us the key in DER format.

Reply:
I guess that it's not a very safe to assume that this will always work. For example, what happens if there's an extra blank line in the beginning of `key:`?

Reply:
what do you recommend? regex for "-----BEGIN CERTIFICATE-----", "-----END CERTIFICATE-----" and whitespace; then remove those?

Reply:
We can use a suitable parsing function that takes care of everything, including whitespace formatting and base64 decoding. That means we have to use a function that decodes PEM (the format including the boundaries) and not DER (a binary representation of the actual key).

```python
return cryptography.hazmat.primitives.serialization.load_pem_public_key(key)
```

However, as there are other places where that applies, too, it may be cleaner to just `return key` here and let the `KeyStore` handle it. Specifically, we can add the call to `load_pem_public_key` in `KeyStore.__init__` and `KeyStore.update`.

As a result, the key will be an object, not a string, so the first argument of `crypto.verify_signature` should become just `pub_key`.